### ISSUE TITLE
Rodi requires __init__ methods defined for Protocols

### ISSUE THREAD

**Issue by akhundMurad** [Neoteroi/rodi#31]
For instance, I have the protocol:
```python
from typing import Protocol, TypeVar, Type

T = TypeVar["T"]

class BuilderProtocol(Protocol[T]):
     def build(self, type_: Type[T]) -> T:
          ....
```
And implementation:
```python
from some_module import Home

class HomeBuilder(BuilderProtocol[Home]):
     def build(self, type_: Type[Home]) -> Home:
          # some logic here
```
Then I am trying to use RODI:
```python
from rodi import Container

container = Container()

container.register(HomeBuilder, BuilderProtocol)

builder = container.resolve(HomeBuilder)
```
There I get exception:
```
rodi.CannotResolveParameterException: Unable to resolve parameter 'args' when resolving 'HomeBuilder'
```

**Comment by akhundMurad** [comment]
Likely, the issue is in [this line](https://github.com/Neoteroi/rodi/blob/main/rodi/__init__.py#L518).
`Singature.from_callable` returns `(self, args, kwargs) -> None` in the case of a generic Protocol class.
Probably, we should add exception [here](https://github.com/Neoteroi/rodi/blob/main/rodi/__init__.py#L480), like the following:
```python
if param_name in ("self", "*args", "**kwargs"):
     continue
```

**Comment by RobertoPrevato** [comment]
Hi @akhundMurad 
You are not registering the types in the right order.

This works:

```python
from rodi import Container
from typing import Generic, TypeVar, Type

T = TypeVar("T")

class BuilderProtocol(Generic[T]):

    def build(self, type_: Type[T]) -> T:
        ...

class Home: pass

class HomeBuilder(BuilderProtocol[Home]):
     def build(self, type_: Type[Home]) -> Home:
          # some logic here
          ...

container = Container()

container.register(BuilderProtocol, HomeBuilder)

builder = container.resolve(BuilderProtocol)

print(builder)
```

```bash
python main.py
<__main__.HomeBuilder object at 0x00000172E4C88650>
````

The idea is that you register a _more generic_ type (like `BuilderProtocol`) to be resolved into a _more concrete_ type so that part of your code can only be concerned / aware of the more generic type.

Example:

```
class CatsProvider:  # generic type, describing the protocol / abstract class
    def get_cats(self) -> list[Cat]:
        ...


class SQLiteCatsProvider(CatsProvider):  # concrete type that interacts with SQLite to store cats information
    ...


class BusinessClassConcernedWithHandlingCats:  # business class that is only concerned with using a CatsProvider and does not care if data is stored in SQLite, PostgreSQL, MongoDB, or any other persistence layer

    cats_provider: CatsProvider


container.register(CatsProvider, SQLiteCatsProvider)
container.register(BusinessClassConcernedWithHandlingCats)


bo = container.resolve(BusinessClassConcernedWithHandlingCats)
assert isinstance(bo.cats_provider, SQLiteCatsProvider)
```

This also makes it possible to plug-in different handlers by app configuration, like in [this example project](https://github.com/Neoteroi/Torino) where 3 kinds of persistence layer are supported.

I hope this helps. 🙏🏼

**Comment by RobertoPrevato** [comment]
Side note because in my first comment I used `Generic` instead of `Protocol`.

The problem here is that `Protocol` generates automatically an `__init__` method that accepts any kind of argument (*args, **kwargs).

`rodi` does not like it because it cannot know how `*args, **kwargs` should be resolved.

So, to use `Protocol`, simply specify an `__init__` method so that `rodi` doesn't have to guess what it should inject when instantiating, like in this example:

```python
from rodi import Container
from typing import Protocol, TypeVar, Type

T = TypeVar("T")

class BuilderProtocol(Protocol[T]):

    def __init__(self) -> None:
         super().__init__()

    def build(self, type_: Type[T]) -> T:
        ...

class Home: pass

class HomeBuilder(BuilderProtocol[Home]):
     def build(self, type_: Type[Home]) -> Home:
          # some logic here
          ...

container = Container()

container.register(BuilderProtocol, HomeBuilder)

builder = container.resolve(BuilderProtocol)

print(builder)
```

I didn't have time to work on a documentation site for `rodi`, I'll document this detail when I get some time.

**Comment by akhundMurad** [comment]
@RobertoPrevato thanks for the explanation, but, in my opinion, it is not good to define `__init__` like this every time

**Comment by akhundMurad** [comment]
My merge request could fix this issue entirely, and you even don't need to define `__init__` with super

**Comment by akhundMurad** [comment]
I am working on the [library](https://github.com/akhundMurad/diator) and want to add support of rodi.
Declining merge request means that users of the diator should add `__init__`  event if they don't need it.
It will give bad user experience to them :(

**Comment by akhundMurad** [comment]
Sure, thanks @RobertoPrevato for your work :handshake:

**Comment by RobertoPrevato** [comment]
Hi @akhundMurad 
Thank You for your contribution. I merged your PR to `main` and I'm releasing `2.0.1` to PyPi tonight.

**Comment by akhundMurad** [comment]
Thank you👍
